### PR TITLE
Fallback for currency snapshot hash in data application traverse

### PR DESCRIPTION
**Pull request by marcinwadon**

### CODE REVIEW

File: modules/currency-l1/src/main/scala/org/tessellation/currency/l1/domain/snapshot/programs/CurrencySnapshotProcessor.scala [hunk 1]
```diff
@@ -224,7 +223,22 @@ object CurrencySnapshotProcessor {
...
+                def usingJson = {
+                  implicit val hasher = Hasher.forJson[F]
+                  snapshot.toHashedWithSignatureCheck
+                }
+
+                usingJson.flatMap {
+                  case Left(_)  => usingKryo
+                  case Right(s) => s.asRight[Signed.InvalidSignatureForHash[CurrencyIncrementalSnapshot]].pure[F]
+                }
+              }).sequence)
```
Comment (sdavidp) [inline modules/currency-l1/src/main/scala/org/tessellation/currency/l1/domain/snapshot/programs/CurrencySnapshotProcessor.scala:241]: Replace the `_.map(_.traverse { snapshot => . . . }).sequence` with `_.traverse(_.traverse { snapshot => . . . })` ?

File: modules/shared/src/main/scala/org/tessellation/currency/dataApplication/DataApplicationTraverse.scala [hunk 2]
```diff
@@ -211,16 +211,29 @@ object DataApplicationTraverse {
...
+                  implicit val hasher = Hasher.forJson[F]
+                  snapshot.toHashedWithSignatureCheck
+                }
+
+                usingJson.flatMap {
+                  case Left(_)  => usingKryo
+                  case Right(s) => s.asRight[Signed.InvalidSignatureForHash[CurrencyIncrementalSnapshot]].pure[F]
+                }
+              }).sequence)
+              .map(_.map(_.traverse(_.toValidatedNel)))
```
Comment (sdavidp) [inline modules/shared/src/main/scala/org/tessellation/currency/dataApplication/DataApplicationTraverse.scala:236]: This code is duplicated. Is it possible to pull it out into common place?
Or maybe generalize it using `EitherT.valueOrF` ?

**Review by sdavidp — APPROVED** [review]
LGTM, added some suggestions

File: modules/shared/src/main/scala/org/tessellation/currency/dataApplication/DataApplicationTraverse.scala [hunk 3]
```diff
@@ -211,16 +211,29 @@ object DataApplicationTraverse {
...
+                def usingJson = {
+                  implicit val hasher = Hasher.forJson[F]
+                  snapshot.toHashedWithSignatureCheck
+                }
+
+                usingJson.flatMap {
+                  case Left(_)  => usingKryo
+                  case Right(s) => s.asRight[Signed.InvalidSignatureForHash[CurrencyIncrementalSnapshot]].pure[F]
+                }
+              }).sequence)
```
Comment (sdavidp) [inline modules/shared/src/main/scala/org/tessellation/currency/dataApplication/DataApplicationTraverse.scala:235]: Maybe something like this?
```suggestion
              .flatMap(_.traverse(_.traverse { snapshot =>
                Hasher.firstRight(Hasher.forJson[F], Hasher.forKryo[F]) { (h: Hasher[F]) =>
                  implicit val ih = h
                  snapshot.toHashedWithSignatureCheck
                }
              }))
```
where we have this
```
def firstRight[F[_]: Monad, A, B](h1: Hasher[F], h2: Hasher[F])(fn: Hasher[F] => F[Either[A, B]]): F[Either[A, B]] =
  fn(h1).flatMap {
    case Left(_) => fn(h2)
    case rb      => rb.pure[F]
  }
```

**Comment by marcinwadon** [comment]
This fix won't be needed

**Pull request closed without merging**